src/librustc/middle/dead.rs

```diff
@@ -557,11 +568,24 @@ impl<'a, 'tcx> Visitor<'tcx> for DeadVisitor<'a, 'tcx> {
     }
 
     fn visit_struct_field(&mut self, field: &'tcx hir::StructField) {
-        if self.should_warn_about_field(&field) {
-            self.warn_dead_code(field.id, field.span,
-                                field.name, "field");
+        if self.need_check_next_union_field {
+            if self.should_warn_about_field(&field) {
+                self.warn_dead_code(field.id, field.span, field.name, "field");
+            } else {
+                let did = self.tcx.hir.get_parent_did(field.id);
+                if let Some(node_id) = self.tcx.hir.as_local_node_id(did) {
+                    match self.tcx.hir.find(node_id) {
+                        Some(hir_map::NodeItem(item)) => match item.node {
+                            // If this is an union's field, it means all previous fields
+                            // have been used as well so no need to check further.
+                            Item_::ItemUnion(_, _) => self.need_check_next_union_field = false,
```

Comment:
The only way I see to mark a field as used is to add the "used" attr into its attr list. However, I can't do it in the `visit_expr` method since the received parameter isn't mutable. Or maybe is there another way I didn't see?